game/natsuki.rpy

```diff
@@ -1,376 +1,677 @@
+                jn_utils.log("Affinity blocked - CN!")
+                return
+
+            if bypass:
+                # Ignore the daily gain and just award the full affinity
+                persistent.affinity += to_add
+                jn_utils.log("Affinity+")
+
+            elif persistent.affinity_daily_gain > 0:
+                # Award the full affinity if any cap remains
+                persistent.affinity_daily_gain -= to_add
+                persistent.affinity += to_add
+
+                if persistent.affinity_daily_gain < 0:
+                    persistent.affinity_daily_gain = 0
+                
+                jn_utils.log("Affinity+")
+
+            else:
+                jn_utils.log("Daily affinity cap reached!")
+
+        @staticmethod
+        def calculatedAffinityLoss(base=1):
+            """
+            Subtracts a calculated amount from affinity, based on the player's relationship with Natsuki.
+
+            IN:
+                - base - The base amount to use for the calculation
+            """
+            persistent.affinity -= base * jn_affinity.get_relationship_length_multiplier()
+            jn_utils.log("Affinity-")
+
+        @staticmethod
+        def percentageAffinityGain(percentage_gain):
+            """
+            Adds a percentage amount to affinity, with the percentage based on the existing affinity value.
+            This bypasses the usual check, so this should only be used for one-off big gains.
+
+            NOTE:
+                If the player has not confessed to Natsuki via the talk_i_love_you topic, further affinity gain is not possible
+                until the confession has been made! This is to prevent players having Natsuki tell them she loves them unwarranted
+                by accidentally crossing the boundary into LOVE.
+
+            IN:
+                - percentage_gain - The integer percentage the affinity should increase by
+            """
+            to_add = persistent.affinity * (float(percentage_gain) / 100)
+            if (not persistent._jn_player_confession_accepted and (persistent.affinity + to_add) > (jn_affinity.AFF_THRESHOLD_LOVE -1)):
+                # Player cannot reach LOVE without having confessed to Natsuki successfully
+                persistent.affinity = jn_affinity.AFF_THRESHOLD_LOVE -1
+                jn_utils.log("Affinity blocked - CN!")
+
+            else:
+                persistent.affinity += to_add
+                jn_utils.log("Affinity+")
+
+        @staticmethod
+        def percentageAffinityLoss(percentage_loss):
+            """
+            Subtracts a percentage amount to affinity, with the percentage based on the existing affinity value.
+
+            IN:
+                - percentage_loss - The integer percentage the affinity should decrease by
+            """
+            persistent.affinity -= persistent.affinity * (float(percentage_loss) / 100)
```

Comment:
I also noticed there's no sanity checks, for example `percentage_loss` could be negative and would result in positive change. Also maybe `persistent.affinity` should be `persistent._jn_affinity`? Also consider encoding aff values in the future.